AGENT-DRAFT: APIM-13008 — JMS Endpoint Plugin (#1182)
gravitee-io-labs/beta-docs: docs/apim/4.10/create-and-configure-apis/configure-v4-apis/endpoints/jms.md

### Overview

> The JMS Endpoint Plugin enables Gravitee APIM to connect to JMS-compliant message brokers as backend endpoints. It supports both producer (send) and consumer (receive) operations across multiple JMS providers including ActiveMQ Classic, ActiveMQ Artemis, IBM MQ, and Solace. This is an Enterprise feature requiring a valid license (pack: `apim-connectors-advanced`, feature ID: `apim-en-endpoint-jms`).

Reviewer: ```suggestion
The JMS Endpoint Plugin enables Gravitee APIM to connect to JMS-compliant message brokers as backend endpoints. It supports both producer (send) and consumer (receive) operations across multiple JMS providers including ActiveMQ Classic, ActiveMQ Artemis, IBM MQ, and Solace. This is an Enterprise feature requiring a valid license (pack: `event-native`, feature ID: `apim-en-endpoint-jms`).
```

#### Connection Sharing

Connection behavior varies by destination type and configuration. Queue consumers always use a shared connection. Topic subscriptions use shared connections for non-durable topics without a client ID, but switch to exclusive connections when a client ID is provided or when durable subscriptions are enabled. For durable topics without an explicit client ID, the plugin auto-generates a UUID.

#### Provider Configuration Models

The plugin supports two configuration models: **provider-specific** (ActiveMQ Classic, ActiveMQ Artemis, IBM MQ, Solace) with dedicated properties, and **generic JNDI-based** for other JMS providers. Provider-specific configurations use direct connection factory instantiation, while the generic model performs JNDI lookups using a configurable properties file format.

### Prerequisites

Before configuring a JMS endpoint, ensure the following requirements are met:

* Gravitee APIM version 4.x or later

> * Valid Enterprise license with `apim-connectors-advanced` pack

Reviewer: ```suggestion
* Valid Enterprise license with `event-native` pack
```

2. Specify the `connectionFactoryClassName` and `apiVersion`.
3. Configure provider-specific properties such as `brokerUrl` for ActiveMQ or `hostname`, `port`, `channel`, and `queueManager` for IBM MQ.
4. Enable producer and/or consumer capabilities in `sharedConfigurationOverride` and specify destination type and name.
5. Add authentication credentials under `security.auth` if required.
6. Deploy the corresponding JMS provider client libraries to `./plugins/jms/ext/` before starting the gateway.

### Configuring Client ID for Topic Subscriptions

1. Set the context attribute `gravitee.attribute.jms.clientId` using the assign-attribute policy to provide an explicit client ID.
2. If no attribute is set, the plugin uses the request client identifier if available.
3. For durable subscriptions without an explicit client ID, the plugin auto-generates a UUID.

Non-durable topic subscriptions without a client ID use a shared connection, while all other scenarios (explicit client ID or durable subscription) use an exclusive connection.

### Restrictions

* Requires Gravitee APIM version 4.x or later

> * Enterprise feature requiring valid license with `apim-connectors-advanced` pack (feature ID: `apim-en-endpoint-jms`)

Reviewer: ```suggestion
* Enterprise feature requiring valid license with `event-native` pack (feature ID: `apim-en-endpoint-jms`)
```
